src/error.rs

```diff
@@ -0,0 +1,43 @@
+use std::result::Result as StdResult;
+
+use thiserror::Error as ThisError;
+
+#[allow(missing_docs)]
+pub type Result<T> = StdResult<T, Error>;
+
+#[allow(missing_docs)]
+#[derive(ThisError, Debug)]
+#[error(transparent)]
+pub struct Error(#[from] InternalError);
+
+macro_rules! impl_from {
+    ($bound:ty) => {
+        impl From<$bound> for Error {
+            fn from(err: $bound) -> Error {
+                InternalError::from(err).into()
+            }
+        }
+    };
+}
+
+impl_from!(log::kv::Error);
+impl_from!(serde_urlencoded::ser::Error);
+impl_from!(serde_urlencoded::de::Error);
+impl_from!(std::io::Error);
+impl_from!(serde_json::Error);
+
+#[derive(ThisError, Debug)]
+pub(crate) enum InternalError {
```

Comment:
Also with `http_types::Error` this shouldn't be necessary, because it wraps the errors and allows you to check [downcasts](https://docs.rs/http-types/2.3.0/http_types/struct.Error.html#method.downcast_ref), like so:
```rust
match surf::get(url).await {
    Ok() -> (),
    Err(error) -> {
        if let Some(&url_err) = error.downcast_ref::<UrlError>() {
            // ...
        } else if let Some(&json_err) = error.downcast_ref::<JsonError>() {
            // ...
        }
    }
}
```